pkg/liqo-controller-manager/webhooks/shadowpod/cache.go

```diff
@@ -0,0 +1,269 @@
+		shadowPod := &shadowPodList.Items[i]
+		nsname := types.NamespacedName{Name: shadowPod.Name, Namespace: shadowPod.Namespace}
+		_, found := pi.shadowPods[nsname.String()]
+		if !found {
+			klog.Warningf("Warning: ShadowPod %s not found in cache", nsname.String())
+			spQuota, _ := getQuotaFromShadowPod(shadowPod, false)
+			pi.addShadowPod(createShadowPodDescription(shadowPod.GetName(), shadowPod.GetNamespace(), shadowPod.GetUID(), spQuota))
+			continue
+		}
+		// Saving a list all the ShadowPods in the cluster that are also in cache
+		spMap[nsname.String()] = struct{}{}
+	}
+	return spMap
+}
+
+func (spv *Validator) checkAlignmentResourceOfferPeeringInfo(ctx context.Context) error {
+	resourceOfferList := sharing.ResourceOfferList{}
+
+	// Get the List of resource offers
+	if err := spv.client.List(ctx, &resourceOfferList, &client.ListOptions{LabelSelector: liqolabels.LocalLabelSelector()}); err != nil {
+		return err
+	}
+
+	// Check if there are new ResourceOffers in the system snapshot
+	for i := range resourceOfferList.Items {
+		ro := &resourceOfferList.Items[i]
+		clusterID := ro.Labels[discovery.ClusterIDLabel]
+		clusterName := retrieveClusterName(ctx, spv.client, clusterID)
+
+		if clusterID == "" {
+			return fmt.Errorf("ResourceOffer for cluster %q has no cluster id", clusterName)
+		}
+
+		// Check if the ResourceOffer is not present in the cache
+		if newPI, found := spv.peeringCache.peeringInfo.LoadOrStore(clusterID, createPeeringInfo(discoveryv1alpha1.ClusterIdentity{
+			ClusterID:   clusterID,
+			ClusterName: clusterName,
+		}, ro.Spec.ResourceQuota.Hard)); !found {
+			klog.V(4).Infof("ResourceOffer %q not found in cache, adding it", clusterName)
+			// Get the List of ShadowPods running on the cluster
+			shadowPodList, err := spv.getShadowPodListByClusterID(ctx, clusterID)
+			if err != nil {
+				return err
+			}
+			newPI.(*peeringInfo).alignExistingShadowPods(shadowPodList)
+		}
+	}
+
+	// Check if PeeringInfos still have corresponding ResourceOffers
+	spv.peeringCache.peeringInfo.Range(
+		func(key, value interface{}) bool {
+			peeringInfo := value.(*peeringInfo)
+			clusterID := key.(string)
+			foundRO := false
+			// Check if the corresponding ResourceOffer is still present in the system snapshot and there are some updates
+			for j := range resourceOfferList.Items {
+				ro := &resourceOfferList.Items[j]
+				if clusterID == ro.Labels[discovery.ClusterIDLabel] {
+					foundRO = true
+					klog.V(5).Infof("Checking for some ResourceOffer Quota updates")
+					peeringInfo.checkAndAlignResourceOfferUpdates(ro.Spec.ResourceQuota.Hard)
+					break
+				}
+			}
+			// If the corresponding ResourceOffer is not present anymore, remove the PeeringInfo from the cache
+			if !foundRO {
+				klog.V(4).Infof("ResourceOffer %q not found in system snapshot, removing it from cache", peeringInfo.clusterIdentity.String())
+				spv.peeringCache.peeringInfo.Delete(clusterID)
+			}
+			return true
+		},
```

Comment:
Given that resource offer quotas are already updated when a new shadowpod is created, this part could be simplified, keeping only the deletion of non-existing peeringInfos.